### PR TITLE
gh-101819: Adapt _io.PyWindowsConsoleIO_Type to heap type

### PR DESCRIPTION
**Pull request by erlend-aasland**
* Issue: gh-101819

### CODE REVIEW

**Comment by vstinner** [comment]
Is it a draft or ready for review?

**Comment by vstinner** [comment]
```
FAIL: test_no_memleak (test.test_embed.MiscTests.test_no_memleak) (frozen_modules='on', stmt='import __hello__')
----------------------------------------------------------------------
Traceback (most recent call last):
  File "D:\a\cpython\cpython\Lib\test\test_embed.py", line 1802, in test_no_memleak
    self.assertEqual(refs, 0, out)
AssertionError: 512 != 0 : [512 refs, 302 blocks]
```

It seems like there is leak somewhere.

**Comment by vstinner** [comment]
```
test_winconsoleio crashed (Exit code 3) -- running: test_multiprocessing_spawn (2 min 47 sec)
Assertion failed: PyObject_TypeCheck(self, (PyTypeObject *)&PyWindowsConsoleIO_Type), file D:\a\cpython\cpython\Modules\_io\winconsoleio.c, line 268
Fatal Python error: Aborted
```

There are still some test failures. Tell me if you need my help to debug that next week.

**Comment by erlend-aasland** [comment]
> Is it a draft or ready for review?

Draft! I don't have a Windows dev environment set up, so I rely on the CI for these changes.

**Comment by erlend-aasland** [comment]
!buildbot Windows.*Refleaks

**Comment by erlend-aasland** [comment]
!buildbot windows.*refleaks

**Comment by erlend-aasland** [comment]
!buildbot windows*refleaks

**Comment by erlend-aasland** [comment]
!buildbot win

**Comment by erlend-aasland** [comment]
!buildbot *win*

**Comment by erlend-aasland** [comment]
!buildbot .*windows.*

**Comment by erlend-aasland** [comment]
!buildbot .*win.*

**Review by kumaraditya303 — APPROVED** [review]

**Comment by erlend-aasland** [comment]
Why isn't the Windows ref leaks bots part of the refleak portion of the fleet?

File: Modules/_io/winconsoleio.c [hunk 1]
```diff
@@ -265,7 +263,10 @@ _io__WindowsConsoleIO___init___impl(winconsoleio *self, PyObject *nameobj,
     int fd_is_own = 0;
     HANDLE handle = NULL;
 
-    assert(PyObject_TypeCheck(self, (PyTypeObject *)&PyWindowsConsoleIO_Type));
+#ifdef Py_DEBUG
```
Comment (vstinner) [inline Modules/_io/winconsoleio.c:266]: You should use ``#ifndef NDEBUG`` instead.

You can build Python in release mode with assertions: https://docs.python.org/dev/using/configure.html#cmdoption-with-assertions

**Review by vstinner — COMMENTED** [review]
Review after merge: LGTM, but the ``#ifdef `` around ``assert()`` is wrong IMO.

File: Modules/_io/winconsoleio.c (hunk 1 above)
Comment (erlend-aasland) [reply Modules/_io/winconsoleio.c:266]: Thanks, I will fix it in a follow-up PR (there will be many before we're done with _io!)

**Pull request merged**